### PR TITLE
Update resorce load

### PR DESCRIPTION
**Pull request by Yankes**
Refactor of handing mod sizes and surface indexes.

Breaking change:
* Mods that try access index outside space allocated for this mod will not load.
* Using negative index is illegal now (except `-1` that always mean "empty").

### CODE REVIEW

File: src/Engine/ModInfo.h [hunk 1]
```diff
@@ -61,10 +62,10 @@ class ModInfo
 	bool canActivate(const std::string &curMaster) const;
 	/// Gets size of mod, bigger mod reserve more values in common colections/surfacesets.
 	int getReservedSpace() const;
-	/// Sets mod size (DO NOT use this method outside Options::updateReservedSpace()).
-	void setReservedSpace(int reservedSpace);
 	/// Gets the OXCE version required by this mod.
 	const std::string &getRequiredExtendedVersion() const;
+	/// Gets ruleset file where are defined based resosruces like required original game data.
```
Comment (SupSuper) [inline src/Engine/ModInfo.h]: resosruces -> resources

File: src/Engine/SoundSet.cpp [hunk 2]
```diff
@@ -190,6 +192,29 @@ Sound *SoundSet::addSound(unsigned int i)
 	return _sounds[i];
 }
 
+/**
+ * Set number of shared sound indexs that are accesable for all mods.
```
Comment (SupSuper) [inline src/Engine/SoundSet.cpp]: accesable -> accessible

File: src/Mod/ExtraSounds.cpp [hunk 3]
```diff
@@ -128,16 +120,28 @@ SoundSet *ExtraSounds::loadSoundSet(SoundSet *set) const
 
 void ExtraSounds::loadSound(SoundSet *set, int index, const std::string &fileName) const
 {
+	int indexWithOffset = index;
+	if (indexWithOffset >= set->getMaxSharedSounds())
+	{
+		if ((size_t)indexWithOffset >= _current->Size)
+		{
+			std::ostringstream err;
+			err << "ExtraSounds '" << _type << "' sound '" << indexWithOffset << "' excess mod '"<< _current->Name <<"' size limit " << _current->Size;
```
Comment (SupSuper) [inline src/Mod/ExtraSounds.cpp]: excess -> exceeds

File: src/Mod/ExtraSprites.cpp [hunk 4]
```diff
@@ -259,43 +248,42 @@ SurfaceSet *ExtraSprites::loadSurfaceSet(SurfaceSet *set)
...
+Surface *ExtraSprites::getFrame(SurfaceSet *set, int index) const
 {
-	Surface *frame = set->getFrame(index);
+	int indexWithOffset = index;
+	if (indexWithOffset >= set->getMaxSharedFrames())
+	{
+		if ((size_t)indexWithOffset >= _current->Size)
+		{
+			std::ostringstream err;
+			err << "ExtraSprites '" << _type << "' frame '" << indexWithOffset << "' excess mod '"<< _current->Name <<"' size limit " << _current->Size;
```
Comment (SupSuper) [inline src/Mod/ExtraSprites.cpp]: excess -> exceeds

File: src/Mod/Mod.h [hunk 5]
```diff
@@ -75,11 +75,27 @@ class RuleGlobe;
...
+class ModInfo;
 class RuleVideo;
 class RuleMusic;
 class RuleMissionScript;
 struct StatAdjustment;
 
+/**
+ * Mod data used when loading resources
+ */
+struct ModData
```
Comment (SupSuper) [inline src/Mod/Mod.h]: Why uppercase variable names here?

File: src/Mod/Mod.cpp [hunk 6]
```diff
@@ -655,7 +656,82 @@ const std::vector<std::vector<Uint8> > *Mod::getLUTs() const
...
 {
-	return _modOffset;
+	return _modCurrent->Offset;
+}
+
+/**
+ * Get offset and index for sound set or sprite set.
+ */
+void Mod::loadOffsetNode(const std::string &parent, int& offset, const YAML::Node &node, size_t shared, size_t multipler) const
+{
```
Comment (SupSuper) [inline src/Mod/Mod.cpp]: Would appreciate documenting all these parameters, like what is "multiplier" for?

File: src/Mod/Mod.cpp [hunk 7]
```diff
@@ -696,73 +795,119 @@ int Mod::getSoundOffset(int sound, const std::string& set) const
...
  * @param mods List of <modId, rulesetFiles> pairs.
  */
 void Mod::loadAll(const std::vector< std::pair< std::string, std::vector<std::string> > > &mods)
 {
 	Log(LOG_INFO) << "Loading rulesets...";
-	std::vector<size_t> modOffsets(mods.size());
+	_modData.clear();
+	_modData.resize(mods.size());
+
+	std::set<std::string> usedModNames;
```
Comment (SupSuper) [inline src/Mod/Mod.cpp]: What are these reserved for? If they're "magic values" for later they should be constants.

File: src/Mod/Mod.cpp [hunk 8]
```diff
@@ -2729,12 +2876,57 @@ void Mod::loadVanillaResources()
...
-	TextButton::soundPress = getSound("GEO.CAT", Mod::BUTTON_PRESS);
-	Window::soundPopup[0] = getSound("GEO.CAT", Mod::WINDOW_POPUP[0]);
-	Window::soundPopup[1] = getSound("GEO.CAT", Mod::WINDOW_POPUP[1]);
-	Window::soundPopup[2] = getSound("GEO.CAT", Mod::WINDOW_POPUP[2]);
-
 	loadBattlescapeResources(); // TODO load this at battlescape start, unload at battlescape end?
+
+	//update number of shared indexes in surface sets and sound sets
+	{
+		std::string surfaceNames[] =
```
Comment (SupSuper) [inline src/Mod/Mod.cpp]: Do these need to be hardcoded? I can already see some missing like INTICON.PCK

File: src/Mod/RuleCraft.cpp [hunk 9]
```diff
@@ -54,7 +54,13 @@ void RuleCraft::load(const YAML::Node &node, Mod *mod, int listOrder)
 	_requires = node["requires"].as< std::vector<std::string> >(_requires);
 	if (node["sprite"])
 	{
-		// this is an offset in BASEBITS.PCK, and two in INTICONS.PCK
+		// used in
+		// Surface set (baseOffset):
+		//   BASEBITS.PCK (33)
+		//   INTICONS.PCK (11)
```
Comment (SupSuper) [inline src/Mod/RuleCraft.cpp]: INTICONS.PCK -> INTICON.PCK

File: src/Mod/Mod.cpp (hunk 8 above)
Comment (Yankes) [reply src/Mod/Mod.cpp]: Overall there are hardcoded in other places. If you want I can add something like `soundDefs` to handle this case too.

**Pull request merged**